### ISSUE TITLE
move reference sequence to top-level metadata

### ISSUE THREAD

**Issue by petrelharp** [MesserLab/SLiM#180]
Currently we store the reference sequence as a separate "column" in the kastore object. This requires us to open up the kastore separately, but more importantly, it requires us to do some gymnastics in pyslim to maintain the reference sequence in the tree sequence. In particular, we need to write a wrapper for any msprime method that modifies a tree sequence that does that operation and then adds back on the reference sequence.

But, now that we've got top-level metadata, we can just put the reference sequence there, and it'll travel with the tree sequence for free. This will remove a lot of responsibility from pyslim, and a lot of potential sources of confusion.

I think the only thing that affects this is whether it should be a standard tskit metadata entry, and hence `ts.metadata["reference_sequence"]`, or whether it should be under the SLiM metadata (and hence `ts.metadata["SLiM"]["reference_sequence"]`. I'd vote for the former.

(*Note:* I've asked for input here on the tskit slack.)

**Comment by bhaller** [comment]
Yes, it seems to make sense to shift this, although it'll be another file version hassle; I guess it's worth it.

The former proposal – `ts.metadata["reference_sequence"]` – makes sense to me.  The only thing I'd add is: probably it would be useful to be able to provide the reference sequence instead as a filename or a URL, and have it read in from that if needed.  We don't want to get into that now, but we ought to make sure that that will fit into the scheme we design now, I suppose.

**Comment by petrelharp** [comment]
Over on the tskit slack, @jeromekelleher said:
```
(1) seems like a good short term solution to me. Right now we're not using the reference sequence in tskit, so it's (by our definition) metadata. When we do, we can think about how to do (3) properly. (By properly, I mean hook up with existing infrastructure so that we don't need to have a copy of GRCh38 in every tsinfer'd tree sequence, but we fetch on demand)
```
... so, let's do this!

> probably it would be useful to be able to provide the reference sequence instead as a filename or a URL, and have it read in from that if needed.

How about making `ts.metadata["reference_sequence"]` an object, and we'll just stick in `ts.metadata["reference_sequence"]["sequence"] = CCGTAGTCCT...`; and that'd leave room in the future for putting in e.g. `ts.metadata["reference_sequence"]["url"] = XXX`?  (Or we just stick the nucleotides straight in to `ts.metadata["reference_sequence"]` and don't prematurely generalize? Do you see SLiM wanting to put in a URL or hash or something?)

**Comment by jeromekelleher** [comment]
I think it's probably better to have ``ts.metadata["reference_sequence"] = "ACGT..."`` rather than making it an object. I think if we do have the fancy on-demand fetching and caching stuff for reference sequences then it'll be done as *data* rather than *metadata* - that is, we use it internally in the library. The idea of metadata is that it's supposed to be opaque to the tskit library, but if start adding custom semantics for certain keys, then that doesn't hold true any more.

So, ultimately, I think we'll have something like ``ts.reference_sequence`` which will have some useful behaviour like automatically fetching and caching the reference for a given (say) human build and contig (we could have some fallback behaviour that checks the metadata for the existence of the ``reference_sequence`` and do something appropriate if it exists).

**Comment by bhaller** [comment]
I'm inclined not to touch this stuff until we are ready to move it into tskit data, rather than metadata.  If I understand correctly, tskit is not yet ready/interested in doing that, but that is the long-term vision.  So, for the time being, "if it ain't broke don't fix it" would be my guiding principle.  :->  Changing it now when we know it's going to change again in future just causes file-version thrash for no reason, which is annoying and increases the likelihood of a bug in the read/write code.  So I'm inclined to close this issue, or at least mark it "will not fix for the foreseeable future" – but simpler to just close it and open a new issue when there's work we actually want to do now.  Thoughts?

**Comment by petrelharp** [comment]
>  So, for the time being, "if it ain't broke don't fix it" would be my guiding principle. :-> 

Agree, but it *is* kinda broken at the moment, because currently it's way too easy for users to lose the reference sequence - to avoid this, pyslim's SlimTreeSequence would need to wrap *all* the tree sequence transformation methods to stick back on the reference sequence. In fact, many of our examples *lose* the reference sequence - any time we wrap something in `SlimTreeSequence( )`, like for instance `SlimTreeSequence(msprime.mutate(ts, ....))`, we have lost it. Moving it to metadata will be much easier than dealing with this, and much less confusing for the users. I don't see an alternative, besides just giving up on the idea of maintaining the reference sequence. It's annoying to have to bump the file version, but we're wanting to do that anyohw (to adjust the population metadata).

**Comment by jeromekelleher** [comment]
I agree with @petrelharp here - sure it's not great doing something that you know will change in the future, but we totally don't have the time to do ``TreeSequence.reference_sequence`` in tskit right now and realistically probably won't for something like a year.

**Comment by bhaller** [comment]
I see.  OK, that provides a pretty strong justification for this.  Hmm.  Can we put it in its final resting place now, and just have tskit not actually do anything with it yet?  Or do we really have to move it to metadata now, and then move it again later?

**Comment by jeromekelleher** [comment]
Tskit only reads and write out columns that it recognises I'm afraid, so ``metadata`` is the only option without changes to tskit.

(That's not to say that I think silently dropping unknown columns is the correct behaviour, but it would be significant work to rectify this.)

**Comment by bhaller** [comment]
OK, I guess I'm convinced.  I guess we put it into the kastore, back in the day, because we didn't have the spiffy metadata stuff yet, at that time?  Well, so, where do we go from here?  One part is that pyslim gets extended to be able to read the sequence from *either* site, and moves it to metadata if it finds it in the kastore, thereby converting old files to an internal format that will maintain the reference sequence across transformations.  And then the other part is that SLiM gets modified to be able to read the sequence from either place, and to always write it out to metadata.  And I guess even if both can read the sequence from either place, we still need to bump the file version, because .trees files that are written out with the sequence in the new place will not be (correctly) readable by older versions of SLiM/pyslim.  Given that, I guess all of these things ought to happen simultaneously.  @petrelharp, do you plan to do the work on the SLiM side, or am I doing that?  And what's your timeframe for this?  It will roll in SLiM 3.7, but at present I have no target date set for that, and I don't want to set one until I really have to; the goal is to roll stuff into it on an ongoing basis until I'm ready to release SLiM 4, ideally.  So it will just be in GitHub until then, which ought to be fine, right?

**Comment by petrelharp** [comment]
>  I guess we put it into the kastore, back in the day, because we didn't have the spiffy metadata stuff yet, at that time?

Right!

>  Well, so, where do we go from here? 

What you say sounds right - should be straightforward: we'll bump the file version, and then if the file version is older than the new one we'll read the reference sequence from the old place, more or less just like we deal with other old file versions. And we can bundle this with the change to population metadata. I'll discuss timelines elsewhere.

**Comment by bhaller** [comment]
tskit is adding support for keeping the reference sequence in kastore at `reference_sequence/data `, as I understand it (see https://github.com/tskit-dev/tskit/issues/146), which is where we used to keep it, so I'm backing out this fix.  I will not change the file version number for this reversal; file version 0.7 was added earlier in the SLiM 3.7 development cycle for the addition of the parents column info in the individuals table, and then this fix was added in to version 0.7, and now I'm going to back it out of 0.7.  Since 0.7 has not gone public yet in a SLiM release, these changes will affect only people on the GitHub head, who ought to be prepared for such flux.  Note that I'm not going to back out the diffs exactly; some restructuring was done for this that remains a good idea, I think.

**Comment by bhaller** [comment]
OK, it has been pointed out by @benjeffery that of course the fix here should use the new tskit APIs to set the reference sequence prior to saving, rather than just pushing it into the kastore.  So, reopening.

**Comment by bhaller** [comment]
Now waiting for https://github.com/tskit-dev/tskit/issues/1970 to be resolved to proceed with this.

**Comment by bhaller** [comment]
This was fixed in #247.
